### DISCUSSION TITLE
Out-of-box Keras 3 Benchmarks: Currently Pytorch is significantly slower than Tensorflow in the training

### DISCUSSION THREAD

**Discussion by jeffDebug** [keras-team/keras#20037] [Q&A]
Referring to the [Keras 3 benchmark page](https://keras.io/getting_started/benchmarks/), it is seen that the PyTorch is significantly slower than the Tensorflow in the training. The following result is extracted from this page to compare the training speed.

### The speed is measured in _ms/step. Lower is better_.
| Training Model | Batch Size | Keras 3 (TensorFlow) | Keras 3 (PyTorch) |
| ------------ | ------------ | ------------ | ------------ |
| SegmentAnything (fit) | 1|  355.25 | 1,388.87 |
| Stable Diffusion (fit) | 8 | 392.24 | 823.44 |
| BERT (fit) | 32 | 214.49 | 808.68 |
| Gemma (fit) | 8 | 232.52 | 525.15 |
| Mistral (fit) | 8 | 185.92 | 452.12 |

There is a reason why we investigated this benchmarking result. For our own DL model, the training speed also became significantly slower (50% to 60% slower), after we migrated from Tensorflow 2 (with GPU/Cuda support) to Keras 3 (with PyTorch+cuda) on Windows. On Windows, the Tensorflow 2.10+ has no GPU/Cuda support anymore, which is one of  reasons why we wanted to switch to Keras 3 with the PyTorch+cuda backend.

In the above benchmarking [page,](https://keras.io/getting_started/benchmarks/) I also noticed the footnote right below the result (Table 2):
> _LLM inference with the PyTorch backend is abnormally slow at this time because KerasNLP uses static sequence padding, unlike HuggingFace. This will be addressed soon._

My question is, in which version of the future release of Keras 3+, will the PyTorch slowness issue be addressed? Should this also generally resolve the training speed issue for the PyTorch backend?

In the current Keras 3, are there any ways I can improve the training speed for the PyTorch backend?

**Comment by jeffDebug** [comment]
Could anyone please give any feedback or suggestion on how to optimize the training for the torch backend?

Referring to [this page](https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/), I see one possible way to try is to use the Cuda Graph. But the underlying torch.nn.Module object must have been wrapped inside the keras.models.Model. I'm wondering if keras 3 provides any API to access underlying torch models to make such optimization possible?

**Comment by abhaskumarsinha** [comment]
Hello Jeff,

To the best of my knowledge, unless PyTorch uses highly optimized linear algebra compilers, CUDA based libs, it is very likely PyTorch to be slower. PyTorch benchmarks often have compiler-aware codes in different LLMs that actually cause that speed up. 

For most of the time JAX backend is usually the fastest for inference and if using `tf.data` dataloaders during training. Feel free to switch the backend to JAX.

Best Regards,
Abhas Kumar Sinha

**Reply by jeffDebug** [reply]
@ghsanti  @abhaskumarsinha

Thanks for the suggestion and reference links.

Concerning the WSL on Windows, we try to avoid adding the WSL (layer) in our product, (1) to minimize the engineering complication when shipping the product, (2) to minimize the runtime overhead and (3) to minimize the size of the product deployment. Therefore, unless it's absolutely necessary, we'd choose not to use the WSL.

Regarding the JAX backend, the GPU/Cuda is not supported on Window yet (see, [this reference link](https://jax.readthedocs.io/en/latest/installation.html#supported-platforms))..

Regarding the tuning the PyTorch backend, so far I have tried the followings.

1. Increasing the _batch size_: this doesn't significantly speed up the training if we keep the same amount of training samples each epoch.
2. Tuning the _steps_per_execution_: I ran into this "_ValueError: `steps_per_execution` must be 1 with the PyTorch backend. Received: steps_per_execution=2_"
3. Enabling _jit_compile=True_: the training crashed into the "_RuntimeError: Cannot find a working triton installation_". I'm not sure if this is the pytorch/triton installation issue, or the compatibility issue between the **_Keras 3.3.3 and torch 2.3.0_**. I'm still trying to figure this out. If you guys have any experience, or advice, I'd be glad to hear about. I assume the flag of "_jit_compile=True_" must have been tested in Keras 3, with some version of PyTorch. I'm wondering which PyTorch version had been tested for this flag.

**Reply by santi-mir** [reply]
Would Docker be less overhead @jeffDebug ? Could allow you to use linux+cuda+tf

For the rest, I can't add much I'm afraid:

1. Yes, that's not surprising
2. I missed this line in the docs:
> **steps_per_execution**: (...) Not supported with the PyTorch backend.


So it has to be **1** indeed.

3. That may need to be raised as an issue unless you can solve it

**Reply by santi-mir** [reply]
You may try this though @jeffDebug ?

> torch.compile is the latest method to speed up your PyTorch code! torch.compile makes PyTorch code run faster by JIT-compiling PyTorch code into optimized kernels, all while requiring minimal code changes.

> torch.compile is included in the latest PyTorch. Running TorchInductor on GPU requires Triton, which is included with the PyTorch 2.0 nightly binary. If Triton is still missing, try installing torchtriton via pip (pip install torchtriton --extra-index-url "https://download.pytorch.org/whl/nightly/cu117" for CUDA 11.7).




[Source](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html)

**Reply by jeffDebug** [reply]
@ghsanti  Much thanks again.

Yes, we do use the docker on the cloud, which is indeed linux+cuda+tf. It is the support of Windows platform that I'm working to resolve for the training performance now.

Yes, I indeed had tried
> pip (pip install torchtriton --extra-index-url

pip wouldn't be able to find the torchtriton for Windows though. As stated in the [triton website](https://github.com/triton-lang/triton?tab=readme-ov-file#compatibility),  it is indeed only the Linux that is supported. I'm still searching around to see if any triton wheel file for Win64 would be available. I may also try playing with different versions of PyTorch, too. So far no resolution is found yet.

**Reply by santi-mir** [reply]
Profiling tools aren't planned atm: https://github.com/keras-team/keras/issues/19476

chollet said:

> There are no profiling plans from our side at this time -- if you want to create this feature, you're welcome to do it, it would be a very valuable addition.

As you may find bit rabbit-holing the thread, in Pytorch you can use [pytorch-lightning](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html) or the [profiler](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html), and in tf there is some [tensorboard app.](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras)
